### ISSUE TITLE
[Enhancement]: Set an option to choose if for non-guest orders to require the user to be logged in before showing thank you page with order details.

### ISSUE THREAD

**Issue by ngrudev** [woothemes/woocommerce#39598]
### Describe the solution you'd like

Currently if non-guest user place an order without logging in, after order is placed there is no thank you page, but a form to log-in the system with a notice "Please log in to your account to view this order.". This is frustrating for me and there is no option to disable this feature. I want to show the thank you page always, no matter if the non-guest user is logged in or not. If this WC function can't be fully disabled for security reasons, there can be a time limit (also set in options), for example 60 minutes, during it the non-guest user to be able to see the thank you page with the order details if not logged-in. After this time expire I am Ok to hide order details if user is not logged in. 

I am using a plugin which converts automatically guest users to registered user after order is placed. For that reason my guest users never see thank you page after checkout, which is confusion for them, and they dont know if the order is placed successfuly or not, neither can review the order details.

The code for this woocommerce behaviour is in file /includes/shortcodes/class-wc-shortcode-checkout.php from line 302 to 319

		// For non-guest orders, require the user to be logged in before showing this page.
		if ( $order_customer_id && get_current_user_id() !== $order_customer_id ) {
			wc_print_notice( esc_html__( 'Please log in to your account to view this order.', 'woocommerce' ), 'notice' );
			woocommerce_login_form( array( 'redirect' => $order->get_checkout_order_received_url() ) );
			return;
		}

		// For guest orders, request they verify their email address (unless we can identify them via the active user session).
		if ( self::guest_should_verify_email( $order, 'order-received' ) ) {
			wc_get_template(
				'checkout/form-verify-email.php',
				array(
					'failed_submission' => ! empty( $_POST['email'] ), // phpcs:ignore WordPress.Security.NonceVerification.Missing
					'verify_url'        => $order->get_checkout_order_received_url(),
				)
			);
			return;
		}

Before executing this code, please add an additional check if the option for executing this code is enabled. In my case I wouldn't use this option at all and will set it disabled. Or you can make is smarter, and set to enable this function, but only after X minutes have passed after order is placed (for example 60 minutes or eny time by users personal choice). During this time the non-guest user can view the thank you page if not logged in, but after this time expire he has to log-in for order details.

Thank you,

### Describe alternatives you've considered

Currently I have commented this code in the class-wc-shortcode-checkout.php , excluding it from this file, but eveytime WC is updated, the file is overwriten and I must comment it again till next update, which is annoying.

```[tasklist]
### Tasks
- [ ] Add new option to disable part of the function
- [ ] Add additional check if the new option is enabled, before executing some parts of the function code
```

**Comment by barryhughes** [comment]
Hi @ngrudev,

> I am using a plugin which converts automatically guest users to registered user after order is placed.

Can you confirm which plugin you are using? WooCommerce natively supports registering new users during checkout and that does not lead to this problem, which makes me wonder if the plugin you are using can be modified slightly?

**Comment by almirsantos99** [comment]
Hi @barryhughes how's it going?

I'm facing the same issue right here, I have a client with a store that have old persons as customers. The store attendant create orders by shopping as customer and them send the link of order-received page to the customer so they can actually pay the order. 

With the first part of the code showed by @ngrudev, **without a filter** is not possible to turn this feature off unless you remove it completely from the WC plugin. 

The second part of the code is possible to solve with the filter woocommerce_order_email_verification_required seted to false.

Can I create a Pull Request so you guys can analysis the solution related to the first part of the code?

PS: The issue also happens on a brand new store if you try to saw the order-received page not logged-in.

**Comment by barryhughes** [comment]
Sorry for the delay, @almirsantos99: requiring verification before showing order details still seems like a very reasonable default, but we also recognize this change has been disruptive. If you'd like to submit a PR, please feel free to select me as a reviewer and we'll take a look.

Something like this is probably reasonable:

- Add a filter allowing safety checks for non-guest orders to be disabled.
- If the filter is 'on', skip next step.
- Otherwise, if the filter is 'off' (which should be the default), perform the existing check which requires the shopper login before viewing the confirmation page.

**Comment by shyam100379** [comment]
@barryhughes Yes, a filter is needed for non-guest orders as well on the checkout success page. Please go through the link @coreymckrill added. 
I understood that Woocommerce want to keep this verification enabled by default. If you could bring some filter, we could use it to disable this verification for non guest orders. 
Another option is a way to set/check X minutes have passed after an order is placed same like we have for guest orders. This would be a more secure way.
Please let us know your thoughts. Thanks in advance.

**Comment by barryhughes** [comment]
I'm not sure there are many more thoughts to share, @shyam100379: I think we're all aligned on the direction and we will be happy to review any PR that is submitted (@almirsantos99 indicated interest in doing so, for instance) or else we will pick this up in due course.

**Comment by ngrudev** [comment]
> Hi @ngrudev,
> 
> > I am using a plugin which converts automatically guest users to registered user after order is placed.
> 
> Can you confirm which plugin you are using? WooCommerce natively supports registering new users during checkout and that does not lead to this problem, which makes me wonder if the plugin you are using can be modified slightly?

Hi, I am using WooCommerce Customers Manager v29.3 by Lagudi Domenico

**Comment by ngrudev** [comment]
> I'm not sure there are many more thoughts to share, @shyam100379: I think we're all aligned on the direction and we will be happy to review any PR that is submitted (@almirsantos99 indicated interest in doing so, for instance) or else we will pick this up in due course.

Thank you for your support. I am happy that the problem is understood and a solution will be sought.

**Comment by Mattgivertz** [comment]
I have implemented the code suggested by @barryhughes as wp-content/mu-plugins/temporary-39750.php i trust that any code changes will nit affect the code we have implemented as it does what is needed

Matt

**Comment by shyam100379** [comment]
@barryhughes I think the solution mentioned in https://github.com/woocommerce/woocommerce/issues/39750 is the same that i  have suggested in the above comments. Its better to bring a grace period for non guest orders as well - that way the security concerns are met and we can have a filter which in turn will help to display thank you page to the customers for a brief amount of time without any verification.

**Comment by shyam100379** [comment]
@barryhughes any updates? Hope the solution will be added in the next release.
Please let us know. Thank you.

**Comment by almirsantos99** [comment]
Hello guys, how's it going? 

Sorry for my late response, tough week. So, I'm going to create the PR probably today and send it so you guys could review it, hope it works and we have that filter on the next minus update. 

I'll send news.

Regards

**Comment by barryhughes** [comment]
For clarity: @almirsantos99's [change](https://github.com/woocommerce/woocommerce/pull/41347) was accepted and merged (thank you for taking the time to submit it!), which is why this issue has been closed. The change itself will become available in our `8.4.0` release.

**Comment by Mattgivertz** [comment]
Will your code solution for 39750 still work as we have implemented it as
an mu plugin and it suits us well.

Regards
Matt

**Comment by barryhughes** [comment]
The (temporary) solution I shared [in this comment](https://github.com/woocommerce/woocommerce/issues/39750#issuecomment-1714375886) should continue to work—I don't anticipate it conflicting with this latest change ✅

**Comment by shyam100379** [comment]
Thank you all.. When we can expect this version? @barryhughes

**Comment by barryhughes** [comment]
We aim for the [2nd Tuesday of each month](https://developer.woo.com/2020/07/22/woocommerce-core-release-calendar/), so I would anticipate `8.4.0` being released on 2023-12-12.

**Comment by shyam100379** [comment]
Guys, Woocommerce released 8.4. I have updated and added the filter: 
`add_filter( 'woocommerce_order_received_verify_known_shoppers', '__return_false' );`
Now i am getting thank you page even if the user is not logged in.

Hurray!

**Comment by ngrudev** [comment]
> Guys, Woocommerce released 8.4. I have updated and added the filter: `add_filter( 'woocommerce_order_received_verify_known_shoppers', '__return_false' );` Now i am getting thank you page even if the user is not logged in.
> 
> Hurray!

I confirm, the filter is working. I see again the Thank you page after purchase, when not logged in. Thank you.
